### PR TITLE
fix: half reduction with multiple sub-iterators

### LINKED ISSUE #74438: cuda low-precision reductions on large tensors produce wrong results
Reductions with low precision inputs (half, bfloat16) that need sub-iterators accumulate directly in output and thus truncate intermediate results:
```.py
import torch
a=torch.ones(2**31, device="cuda", dtype=torch.half)
a[2**30:]=-1
print(a.sum(), a.float().sum()) #inf and 0, should be 0 and 0
print(a.mean(), a.float().mean()) #inf and 0, should be 0 and 0
```
It's related to #15837, but when #15837 was fixed it incorrectly assumed that reduction directly in output can be done for the case above. 

cc @ezyang @gchanan @zou3519 @ngimel

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #74438 

TODO:
* [x] Add test

### CODE REVIEW

**Comment by kshitij12345** [comment]
@ngimel does this make sense? Thanks.

File: test/test_reductions.py [hunk 1]
```diff
@@ -3373,6 +3373,15 @@ def to_numpy(input):
 
             self.assertEqual(actual, expected, msg, exact_dtype=exact_dtype)
 
+    @onlyCUDA
+    @largeTensorTest("8GB")
+    @dtypes(torch.half, torch.chalf, torch.bfloat16)
```
Comment (kshitij12345) [inline test/test_reductions.py:3378]: Bfloat16 does not suffer from this as it has the range same as float.

File: aten/src/ATen/native/cuda/Reduce.cuh [hunk 2]
```diff
@@ -1135,8 +1135,16 @@ inline void gpu_reduce_kernel(TensorIterator& iter, const ops_t& ops, ident_t id
 
   using traits = function_traits<decltype(&ops_t::reduce)>;
   using arg_t = typename traits::template arg<0>::type;
+  // at::Half/at::ComplexHalf overflows easily as it's range is very small.
+  // So when scalar_t and out_scalar_t are at::Half/at::ComplexHalf, we
+  // set can_accumulate_in_output to False.
+  static constexpr bool is_inp_out_type_half_or_chalf =
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Reduce.cuh:1141]: this should include bfloat also. Even though bfloat range is enough, the intermediate truncation will happen.

Comment (kshitij12345) [reply aten/src/ATen/native/cuda/Reduce.cuh:1141]: Have addressed the review :)

**Review by ngimel — APPROVED** [review]

File: test/test_reductions.py [hunk 3]
```diff
@@ -3382,6 +3383,23 @@ def to_numpy(input):
...
```
Comment (malfet) [inline test/test_reductions.py:3388]: Well, this test fails on large GPUs with expected "mean_cuda is not implemented for 'ComplexHalf'"

Comment (kshitij12345) [reply test/test_reductions.py:3388]: I see that skip has been added. Thanks for checking and adding skip.

**Pull request closed without merging**